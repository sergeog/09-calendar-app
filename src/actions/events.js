import Swal from 'sweetalert2';

import { fetchWithToken } from '../helpers/fetch';
import { prepareEvents } from '../helpers/prepareEvents';
import { types } from '../types/types';

export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;
    try {
      const response = await fetchWithToken('events', event, 'POST');
      const body = await response.json();
      if (body.ok) {
        event.id = body.event.id;
        event.user = { _id: uid, name };

        dispatch(eventAddNew(event));
      }
      console.log(event);
    } catch (error) {
      console.log(error);
    }
  };
};

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const response = await fetchWithToken('events');
      const body = await response.json();
      const events = prepareEvents(body.events);
      dispatch(eventsLoaded(events));
    } catch (error) {
      console.log(error);
    }
  };
};

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithToken(`events/${event.id}`, event, 'PUT');
      const body = await response.json();
      if (body.ok) {
        dispatch(eventUpdated(event));
      } else {
        console.log('body', body);
        Swal.fire('Error', body.message, 'error');
      }
    } catch (error) {
      console.log(error);
    }
  };
};

export const eventStartDelete = () => {
  return async (dispatch, getState) => {
    const { id } = getState().calendar.activeEvent;
    try {
      const response = await fetchWithToken(`events/${id}`, {}, 'DELETE');
      const body = await response.json();
      if (body.ok) {
        dispatch(eventDeleted());
      } else {
        Swal.fire('Error', body.message, 'error');
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventAddNew = (event) => ({
  type: types.EVENT_ADD_NEW,
  payload: event,
});

export const eventSetActive = (event) => ({
  type: types.EVENT_SET_ACTIVE,
  payload: event,
});

export const eventClearActive = () => ({ type: types.EVENT_CLEAR_ACTIVE });

const eventUpdated = (event) => ({
  type: types.EVENT_UPDATED,
  payload: event,
});

const eventDeleted = () => ({ type: types.EVENT_DELETED });

const eventsLoaded = (events) => ({
  type: types.EVENT_LOADED,
  payload: events,
});

export const eventLogout = () => ({
  type: types.EVENT_LOGOUT,
});
