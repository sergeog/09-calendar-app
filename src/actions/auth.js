import Swal from 'sweetalert2';

import { fetchWithoutToken, fetchWithToken } from '../helpers/fetch';
import { types } from '../types/types';
import { eventLogout } from './events';

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const response = await fetchWithoutToken(
      'auth',
      { email, password },
      'POST'
    );

    const body = await response.json();
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(
        login({
          uid: body.uid,
          name: body.name,
        })
      );
    } else {
      Swal.fire('Error', body.message, 'error');
    }
  };
};

export const startRegister = (email, password, name) => {
  return async (dispatch) => {
    const response = await fetchWithoutToken(
      'auth/new',
      { name, email, password },
      'POST'
    );

    const body = await response.json();
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(
        login({
          uid: body.uid,
          name: body.name,
        })
      );
    } else {
      Swal.fire('Error', body.message, 'error');
    }
  };
};

export const startChecking = () => {
  return async (dispatch) => {
    const response = await fetchWithToken('auth/renew');
    const body = await response.json();

    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(
        login({
          uid: body.uid,
          name: body.name,
        })
      );
    } else {
      dispatch(checkingFinish());
    }
  };
};

const checkingFinish = () => ({ type: types.AUTH_CHECKING_FINISH });

const login = (user) => ({
  type: types.AUTH_LOGIN,
  payload: user,
});

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(eventLogout());
    dispatch(logout());
  };
};

export const logout = () => ({ type: types.AUTH_LOGOUT });
