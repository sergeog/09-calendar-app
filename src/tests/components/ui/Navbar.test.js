import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { Navbar } from '../../../components/ui/Navbar';
import { startLogout } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
  startLogout: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {
  auth: {
    name: 'Serge',
  },
};
const store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <Navbar />
  </Provider>
);

describe('Pruebas en <Navbar />', () => {
  test('debe mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe hacer click en Salir', () => {
    wrapper.find('button').simulate('click');
    expect(startLogout).toHaveBeenCalled();
  });
});
