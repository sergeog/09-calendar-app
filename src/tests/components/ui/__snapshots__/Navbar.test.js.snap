// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Navbar /> debe mostrarse correctamente 1`] = `
<div
  className="navbar navbar-dark bg-dark mb-4"
>
  <span
    className="navbar-brand"
  >
    Serge
  </span>
  <button
    className="btn btn-outline-danger"
    onClick={[Function]}
  >
    <i
      className="fas fa-sign-out-alt"
    />
    <span>
       Salir
    </span>
  </button>
</div>
`;
